src/HamiltonianBuilder/FrozenCore.cpp

```diff
@@ -0,0 +1,195 @@
+ *  @param hamiltonian_parameters       the Hamiltonian parameters in an orthonormal orbital basis
+ *  @param x                            the vector upon which the Hamiltonian acts
+ *  @param diagonal                     the diagonal of the Hamiltonian matrix
+ *
+ *  @return the action of the Hamiltonian on the coefficient vector
+ */
+Eigen::VectorXd FrozenCore::matrixVectorProduct(const HamiltonianParameters& hamiltonian_parameters, const Eigen::VectorXd& x, const Eigen::VectorXd& diagonal) const {
+
+    // Freeze Hamiltonian parameters
+    HamiltonianParameters frozen_ham_par = this->freezeHamiltonianParameters(hamiltonian_parameters, X);
+
+    // perform matvec through conventional CI
+    return this->hamiltonian_builder->matrixVectorProduct(frozen_ham_par, x, diagonal);
+}
+
+
+/**
+ *  @param hamiltonian_parameters       the Hamiltonian parameters in an orthonormal orbital basis
+ *
+ *  @return the diagonal of the matrix representation of the Hamiltonian
+ */
+Eigen::VectorXd FrozenCore::calculateDiagonal(const HamiltonianParameters& hamiltonian_parameters) const {
+
+    // Freeze Hamiltonian parameters
+    HamiltonianParameters frozen_ham_par = this->freezeHamiltonianParameters(hamiltonian_parameters, X);
+
+    // calculate diagonal through conventional CI
+    Eigen::VectorXd diagonal = this->hamiltonian_builder->calculateDiagonal(frozen_ham_par);
+
+    // diagonal correction
+    double value = this->diagonalValue(hamiltonian_parameters, X);
+    diagonal += (value * Eigen::VectorXd::Ones(this->get_fock_space()->get_dimension()));
+
+    return diagonal;
+}
+
+
+
+/*
+ *  PUBLIC METHODS
+ */
+
+/**
+ *  @param hamiltonian_parameters               the Hamiltonian parameters in an orthonormal orbital basis
+ *  @param X                                    amount of frozen core orbitals
+ *
+ *  @return new set of Hamiltonian parameters modified to fit the frozen core CI algorithm
+ */
+HamiltonianParameters FrozenCore::freezeHamiltonianParameters(const HamiltonianParameters &hamiltonian_parameters, size_t X) const {
+
+    size_t K = hamiltonian_parameters.get_K();
+    size_t Kn = K - X;  // non-frozen orbitals
+
+    std::shared_ptr<AOBasis> ao_basis;  // nullptr
+
+    // copy one electron integrals from the non-frozen orbitals
+    OneElectronOperator S (hamiltonian_parameters.get_S().get_matrix_representation().block(X, X, Kn, Kn));
+    Eigen::MatrixXd h (hamiltonian_parameters.get_h().get_matrix_representation().block(X, X, Kn, Kn));
+
+    const auto& g = hamiltonian_parameters.get_g();
+
+    // copy two electron integrals from the non-frozen orbitals
+    Eigen::Tensor<double, 4> g_SO = tensorBlockCreation(g.get_matrix_representation(), X, X, X, X);
+
+    // Modify one electron integrals with the missing frozen orbital two electron integral evaluations according to the frozen core CI algorithm
+    for (int i = 0; i < Kn; i++) {
+
+        size_t q = i + X;  // map index to the non-modified Hamiltonian parameters
```

Comment:
use better documentation "map index to the non-modified Hamiltonian parameters" is too convoluted